boot/bootutil/src/swap_scratch.c

```diff
@@ -514,6 +515,17 @@ boot_swap_sectors(int idx, uint32_t sz, struct boot_loader_state *state,
      * controls if special handling is needed (swapping last sector).
      */
     last_sector = boot_img_num_sectors(state, BOOT_PRIMARY_SLOT) - 1;
+    sector_sz = boot_img_sector_size(state, BOOT_PRIMARY_SLOT, 0);
```

Comment:
sectors might be of different size each other,

Reply:
But does MCUboot support different sectors sizes within a single slot?

Reply:
Yes it does: https://github.com/mcu-tools/mcuboot/blob/main/sim/src/image.rs#L342

Reply:
Correct me if I am wrong, but `FlashId::Image0` starts af offset **0x20000** which already falls into the sector with 128KB size.

Reply:
> Correct me if I am wrong, but `FlashId::Image0` starts af offset **0x20000** which already falls into the sector with 128KB size.

Yeah, you're right.

Reply:
Just tried here and it works as expected:

```diff
diff --git a/sim/src/image.rs b/sim/src/image.rs
index b45791e3..6e771871 100644
--- a/sim/src/image.rs
+++ b/sim/src/image.rs
@@ -341,7 +341,9 @@ impl ImagesBuilder {
                 // STM style flash.  Large sectors, with a large scratch area.
                 let dev = SimFlash::new(vec![16 * 1024, 16 * 1024, 16 * 1024, 16 * 1024,
                                         64 * 1024,
-                                        128 * 1024, 128 * 1024, 128 * 1024],
+                                        32 * 1024, 32 * 1024, 64 * 1024, // primary
+                                        32 * 1024, 32 * 1024, 64 * 1024, // secondary
+                                        128 * 1024],                     // scratch
                                         align as usize, erased_val);
                 let dev_id = 0;
                 let mut areadesc = AreaDesc::new();
```

Reply:
Now thinking better about it, it was good point you made, I think it's better to have this memory model above in the tests, so I will submit this patch once I am finished with "professional" work for the day.